File: SWEA/Lv.D3/1206_View.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int main(int argc, char** argv)
{
	int test_case;
	int T;
	T = 10;
	/*
	   여러 개의 테스트 케이스가 주어지므로, 각각을 처리합니다.
	*/
	for(test_case = 1; test_case <= T; ++test_case)
	{
        int N;
        int height[1000];
        int result = 0;
        //입력
        cin >> N;
        for (int i=0; i<N; ++i){
            cin >> height[i];
        }
        //각 빌딩마다 탐색
        for (int i = 0; i < N-4 ; ++i){
            vector<int>semi_height(height+i, height+i+5);
            int building = semi_height[2];
            sort(semi_height.begin(), semi_height.end());
            if (semi_height[4] == building){
                result += (building - semi_height[3]);
        	}
        }
       cout << "#" << test_case << " "<< result << endl;
    }
    return 0;
}



//2. 코드 최적화 풀이

#include <cstdio>
 
int max_v(int a, int b) { return a>b?a:b;}
 
int gm(int left, int right, int building_index, int height[], int &N){
    int maximum = 0;
    for (int i = left; i < right; i++){
        if (i == building_index) continue;
        maximum = max_v(maximum, height[i]); }
    return maximum; }
     
    int main (){
        int T = 10, test_case = 1, answer, N, maximum_height;
        int height[1000];
         
        while (T--){
            answer = 0 ; 
            //입력
            scanf("%d", &N);
            for (int i=0; i<N; i++){
                scanf("%d", &height[i]);}
             
            //빌딩 순회
            for (int i=2;i<N-2;i++){
                maximum_height = gm(i-2, (i + 3 < N) ? i + 3 : N, i, height, N); // i+3 이 N을 넘어가지 않도록 처리
                if (maximum_height < height[i])
                    answer += height[i] - maximum_height;
                }
            printf("#%d %d\n",  test_case++,  answer);
        }
        return 0;
    }

File: SWEA/Lv.D3/1206_View.py
T = 10
# 여러개의 테스트 케이스가 주어지므로, 각각을 처리합니다.
for test_case in range(1, T+1):
    N = int(input())
    height = list(map(int, input().split()))
    result = 0 
    for i in range (len(height)-4):
        semi_height = height[i:i+5]
        building = semi_height[2] # 가운데 빌딩 높이를 기준으로 해서 비교
        semi_height.sort()
        if (max(semi_height) == building): # 가운데 빌딩이 가장 높다면
            result += (building - semi_height[3]) # 두 번째로 높은 빌딩과의 차이가 조망권이 확보된 세대의 수
    print("#{} {}".format(test_case,result))

# 2. 코드 최적화 풀이

for test_case in range(1,11):
    result = 0
    houseCount = int(input())
    house = list(map(int , input().split()))
    for i in range(2, houseCount-2):
        def_2 = house[i] - house[i-2]
        def_1 = house[i] - house[i-1]
        def1 = house[i] - house[i+1]
        def2 = house[i] - house[i+2]
        if def_2 > 0 and def_1 > 0 and def1 > 0 and def2 > 0 :
            result += min(def_2, def_1, def1, def2)
 
    print("#{} {}".format(test_case,result))
